File: Answer/112.路径总和.cs
/*
 * @lc app=leetcode.cn id=112 lang=csharp
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool getSum(TreeNode node, int sum, int ans)
    {
        ans += node.val;
        if(node.left != null && node.right != null)
        {
            return (getSum(node.left, sum, ans) || getSum(node.right, sum, ans));
        }
        else if(node.left != null)
        {
            return getSum(node.left, sum, ans);
        }
        else if(node.right != null)
        {
            return getSum(node.right, sum, ans);
        }
        else
        {
            return sum == ans;
        }
    }

    public bool HasPathSum(TreeNode root, int sum) {
        if(root == null) return false;
        return getSum(root, sum, 0);
    }
}
// @lc code=end

File: 112.路径总和.cpp
/*
 * @lc app=leetcode.cn id=112 lang=cpp
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	bool searchTree(TreeNode* t, int d, int sum)
	{
		if (t->left == nullptr && t->right == nullptr)
		{
			if (d == sum) return true;
			return false;
		}
		if (t->left != nullptr)
		{
			if (searchTree(t->left, d + t->left->val, sum)) return true;
		}
		if (t->right != nullptr)
		{
			if (searchTree(t->right, d + t->right->val, sum)) return true;
		}
		return false;
	}
	
    bool hasPathSum(TreeNode* root, int sum) {
		if (root == nullptr) return false;
		return searchTree(root, root->val, sum);
    }
};
// @lc code=end

